brpc-java-core/src/main/java/com/baidu/brpc/protocol/push/impl/DefaultServerPushProtocol.java

```diff
@@ -0,0 +1,350 @@
+        DefaultSPHead spHead = (DefaultSPHead) response.getSpHead();
+        if (spHead == null) {
+            spHead = new DefaultSPHead((int) response.getLogId(), bodyBytes.length);
+        } else {
+            spHead.bodyLength = bodyBytes.length;
+        }
+        spHead.logId = response.getLogId();
+        spHead.type = SPHead.TYPE_RESPONSE;
+        byte[] headBytes = headToBytes(spHead);
+        return Unpooled.wrappedBuffer(headBytes, bodyBytes);
+    }
+
+    @Override
+    public void afterResponseSent(Request request, Response response, ChannelFuture channelFuture) {
+
+    }
+
+    @Override
+    public Request decodeRequest(Object packet) throws Exception {
+        Request request = this.createRequest();
+        DefaultServerPushPacket spPacket = (DefaultServerPushPacket) packet;
+        SPBody spBody = spPacket.getSpBody();
+        request.setLogId(spPacket.getSpHead().getLogId());
+        request.setSpHead(((DefaultServerPushPacket) packet).getSpHead());
+        decodeRequestBody((spPacket).getSpBody(), request);
+        //request.setTarget(rpcMethodInfo.getTarget());
+        request.setArgs(spBody.getParameters());
+        request.setMethodName(spBody.getMethodName());
+        request.setServiceName(spBody.getServiceName());
+        //  request.setTargetMethod(rpcMethodInfo.getMethod());
+        return request;
+    }
+
+    @Override
+    public boolean returnChannelBeforeResponse() {
+        return false;
+    }
+
+    @Override
+    public DefaultServerPushPacket decode(ChannelHandlerContext ctx, DynamicCompositeByteBuf in,
+                                          boolean isDecodingRequest)
+            throws BadSchemaException, TooBigDataException, NotEnoughDataException {
+        if (in.readableBytes() < DefaultSPHead.SPHEAD_LENGTH) {
+            throw new NotEnoughDataException();
+        }
+        DefaultServerPushPacket packet = new DefaultServerPushPacket();
+        ByteBuf fixHeaderBuf = in.retainedSlice(DefaultSPHead.SPHEAD_LENGTH);
+        try {
+
+            DefaultSPHead spHead = headFromByteBuf(fixHeaderBuf);
+            packet.setSpHead(spHead);
+            int bodyLength = spHead.bodyLength;
+
+            // 512M
+            if (bodyLength > 512 * 1024 * 1024) {
+                throw new TooBigDataException("to big body size:" + bodyLength);
+            }
+
+            if (in.readableBytes() < DefaultSPHead.SPHEAD_LENGTH + bodyLength) {
+                throw new NotEnoughDataException();
+            }
+
+            in.skipBytes(DefaultSPHead.SPHEAD_LENGTH);
+            ByteBuf bodyBuf = in.readRetainedSlice(bodyLength);
+            int readableBytes = bodyBuf.readableBytes();
+            byte[] bodyBytes = new byte[readableBytes];
+            bodyBuf.readBytes(bodyBytes);
+            Schema<SPBody> schema = RuntimeSchema.getSchema(SPBody.class);
+            SPBody spBody = new SPBody();
+            ProtobufIOUtil.mergeFrom((byte[]) bodyBytes, spBody, schema);
```

Comment:
body decode logic should be moved to work thread pool, instead of io thread pool.